Make protobuf dependency optional

fralalonde opened the pull request:
Following #210, I've made the `protobuf` dependency optional. It's still enabled by default but it can be turned off by specifying empty `default-features`. 

Because I also believe that `protobuf` should no longer be enabled by default, I've marked the module as `#![deprecated]`. This is a transition measure before taking out the `protobuf` feature out of the crate's default features. I'll leave the final verdict on this to the maintainers.

How it works: Since the protobuf module code defines some core structs, I've made an alternative copy called `no_proto.rs` which does not impl  `protobuf` trait. Regular `Option` and `Vec` are used in place of `SingularField` and `RepeatingField`, protobuf-specific extra fields were removed and a `from_vec!` macro was introduced.

The non-protobuf binaries are roughly half the size of the original ones and compile about 33% faster:
```
cargo build --release --all
ls -la target/release
[...]
-rw-r--r--  2 fralalonde fralalonde 3539952 Dec 28 07:05 libprometheus.rlib
-rw-r--r--  1 fralalonde fralalonde     419 Dec 28 06:58 libprometheus_static_metric.d
-rwxr-xr-x  2 fralalonde fralalonde 2838384 Dec 28 06:58 libprometheus_static_metric.so
```

```
cargo build --release --all --no-default-features
ls -la target/release
[...]
-rw-r--r--  2 fralalonde fralalonde 1779236 Dec 28 07:06 libprometheus.rlib
-rw-r--r--  1 fralalonde fralalonde     419 Dec 28 06:58 libprometheus_static_metric.d
-rwxr-xr-x  2 fralalonde fralalonde 2838384 Dec 28 06:58 libprometheus_static_metric.so
```

Although `cargo test` passes without problem, I haven't tested the new binaries against an actual Prometheus server. I think this should be done before merging since this is a sizable change but I am not equipped to run this re-qualification step.

breeswish commented:
Thanks! This is a large PR, I need some time to take a look.

fralalonde commented:
Cool! You'll find that all methods implemented in the new `no_proto.rs` match existing methods in the original `proto/metrics.rs`. I thought we should also change the actual module name to something else than `mod proto`, but I couldnt think of a good name. Maybe it's ok to just wait a few versions when protobuf is completely removed from the code.

siddontang commented:
PTAL @brson

siddontang commented on Cargo.toml:
```diff
@@ -14,14 +14,14 @@ documentation = "https://docs.rs/prometheus"
 
 [features]
-default = []
+default = ["protobuf"]
```
seem we can make protobuf not default now. 
@breeswish

brson replied:
Makes sense to me.

breeswish commented:
Thanks! What I mean is that, the best way is to write a foundation (e.g. these support structures) that does not have an idea of the protobuf totally from the base. Then we adapt the protobuf over this foundation. Don't be afraid of making it looks differently compared to the protobuf. In fact it is the purpose. For example, we can build much more rustic and easier to use interface when these structures are crafted by our own instead of protobuf. Now, protobuf is optional and we are indeed crafting these structures by our own, no longer restricted by the limited functionality generated by protobuf, then just do it :)

brson commented on src/errors.rs:
```diff
@@ -41,6 +43,7 @@ quick_error! {
             description(err.description())
             display("Io {}", err)
         }
+
```
Please remove this extra space.

brson commented on src/errors.rs:
```diff
@@ -51,5 +54,34 @@ quick_error! {
+            display("Io {}", err)
+        }
+    }
+}
```
This duplication of the entire error declaration is quite unfortunate. I can't think how to avoid it offhand though :-/

brson commented:
Not being to familiar with this codebase, it looks to me like the amount of duplication between metrics.rs and no_proto.rs could be a big maintenance problem if there is no code that is checking for a 1:1 correspondence in the definitions of the two. Any changes in metrics.rs need to be reflected in no_proto.rs and vice-versa.

Perhaps along the lines of what @breeswish is saying, one way to avoid that problem would be to wedge another layer between these two 'proto' definitions and the library, such that every single type and function in the proto files must be defined similarly for the crate to compile, e.g.:

```
// These two are implementation details
#[cfg(feature = "protobuf")]
mod proto_impl {
    impl LabelPair {
        pub fn clear_name() { ... }
    }
}

#[cfg(feature = "protobuf")]
mod proto_impl {
    impl LabelPair {
        pub fn clear_name() { ... }
    }
}

// This is the actual interface
mod proto {
    use super::proto_impl;

    struct LabelPair(proto_impl::LabelPair);

    impl LabelPair {
        pub fn clear_name() {
            self.0.clear_name();
        }
    }
}
```

That way there is a clear single interface to these definitions, and if either of the implementations drifts in their definitions, it will be known at compile time. This also creates an interface that is under our own control, not determined by protobuf, as I think @breeswish was implying would be useful.

@breeswish what do you think of that?

I'm also a bit concerned about how the cfg's are 'scattered' around the codebase, which can make it difficult to know if one is making the correct changes for both configurations during maintenance. I have to head out right now, but I'll circle back and think if I can see any ways to reduce the number of protobuf cfgs.

Awesome patch @fralalonde. Sorry for the delay in reviewing.

fralalonde commented:
I've minimized the surface of the impls according to their current usage in the code, removing many unused protobuf-generated fns. This is not the isolation mechanism that is sought but will reduce the work in building it.

I've also remamed the `proto` module to `model` which I believe suits it's purpose better. The pbuf-generated version is now called `proto-model` while the non proto is called `plain_model`. I've tested & fixed minor issues with code generation. 

I am unfamilar with the CI process but I thought there should now be two separate test builds , a non-proto build using `--no-default-features`and the existing one implying `--features="protobuf"` to validate both variants of the crate.

brson commented:
> I am unfamilar with the CI process but I thought there should now be two separate test builds , a non-proto build using `--no-default-features`and the existing one implying `--features="protobuf"` to validate both variants of the crate.

Yes, I agree. It looks like there are a few different ways that might be accomplished, but just adding another line like [this](https://github.com/pingcap/rust-prometheus/blob/master/.travis.yml#L38) and adjusting the `--features` flag seems like it would work fine.

@fralalonde I like the name 'model' better than 'proto' but I'm less confident about the change to delete API surface from `plain_model`; and the reason is that `model` / `proto` is a public export in lib.rs - it's part of the public API. So deleting API from `plain_model` will make the public API different based on whether protobufs is on or not. If this API wasn't public then it wouldn't matter much if they were consistent. Maybe we can here from @breeswish whether that public export is needed.

Renaming the `proto` export to `model` requires a version bump. As would changing the default feature set.

I'll wait to here from @breeswish before suggesting anything concrete, because I am not familiar with this crate.

fralalonde commented:
@brson You make a good point regarding the variable geometry of API and mod name change compatibility breakage. The `model` mod should be introduced while just deprecating the `proto` mod. `model` would define traits with only common methods exposed. 

If there was going to be a version bump because of API change, it might not be worth keeping protobuf at all, IMO. But if we're going to keep protobuf at all, there needs to be a smooth mod-upgrade / protobuf-exit path. If you allow me, I'll put in some more work in.

brson commented:
@fralalonde please be my guest.

fralalonde commented:
I've renamed the module name `model` back to `proto` for backward compatibility, but both protobuf and plain variations keep the `*_model.rs` file names. Having module aliases would allow progressive introduction / deprecation of the new module. The unused-but-exposed methods now get a `#[deprecated]` attribute instead.

I've also put back  _most_ of the unused but exposed methods that I removed. Some of the original protobuf-generated methods can no longer be added back because the new model no longer `Option`-wraps everything as protobuf did, leading to methods `has_*` and `clear_*` no longer making sense.

breeswish commented:
@overvenus PTAL

Hoverbear commented on src/plain_model.rs:
```diff
@@ -0,0 +1,411 @@
+}
+
+impl LabelPair {
+    #[deprecated(note = "Use default()", since = "0.5.1")]
```
I don't think this is necessary. It's less "deprecated" and more "the function signature may change over time"

fralalonde replied:
Hm, I don't think `new()`'s signature would change if we kept it? 

I deprecated all the `new()` because all they do is call `default()` which is redundant. I wanted to get rid of to protobuf-isms and take the API closer to what it would be like if it hadn't been generated.

fralalonde replied:
I'm ok with not deprecating stuff, this can be done later. What would be your suggestion?

Hoverbear commented:
Thanks so much for this PR! :)
